PostgreSQL JSON fields

signedav opened the pull request:
Viewing of json/jsonb type fields of PostgreSQL possible.

Display map values as "Key/Value" or array values as "List":
![listandmap](https://user-images.githubusercontent.com/28384354/45422344-3bd28b80-b690-11e8-88b1-952bc22111aa.png)

Display values as "Text":
![text](https://user-images.githubusercontent.com/28384354/45422363-42f99980-b690-11e8-83d3-a9ef44f5be4f.png)

Of course its not possible to display JSON maps as "List" and arrays as "Key/Value".

Storing data is not yet supported.

signedav committed:
json/jsonb type integration

an tests for map and list

signedav committed:
test db stuff json

signedav committed:
displayString on textoutput

but only works if json contains a map...

and codestyle in python

signedav committed:
displaystring of every json object

Removed unused stuff

nirvn commented:
Oh yes please! It'd be nice if that made its way into 3.4.

signedav committed:
removed unused docs

and unused db entry

nyalldawson commented on src/core/qgsfield.cpp:
```diff
@@ -253,6 +254,11 @@ QString QgsField::displayString( const QVariant &v ) const
     if ( ok )
       return QLocale().toString( converted );
   }
+  else if ( d->typeName.compare( "json" ) == 0 || d->typeName.compare( "jsonb" ) == 0 )
```
QLatin1String

nyalldawson commented on src/gui/editorwidgets/qgslistwidgetfactory.cpp:
```diff
@@ -44,5 +44,5 @@ QgsEditorConfigWidget *QgsListWidgetFactory::configWidget( QgsVectorLayer *vl, i
 {
   const QgsField field = vl->fields().field( fieldIdx );
-  return ( field.type() == QVariant::List || field.type() == QVariant::StringList ) && field.subType() != QVariant::Invalid ? 20 : 0;
+  return ( field.type() == QVariant::List || field.type() == QVariant::StringList || field.typeName().compare( QLatin1String( "json" ) ) == 0 || field.typeName().compare( QLatin1String( "jsonb" ) ) == 0 ) && field.subType() != QVariant::Invalid ? 20 : 0;
```
Just wondering, should this be case insenstive?

signedav replied:
I should never be uppercase. But I can add it to be sure here.

signedav replied:
I took the `==` that is sensitive, because it should be always lowercase, and so it's the same comparison like it's done on other types and there is no confusion with `compare`. In case it should be case intensive, we should change all the comparisons. I could do that, but I don't see importance here...

nyalldawson commented on src/providers/postgres/qgspostgresprovider.cpp:
```diff
@@ -230,11 +230,13 @@ QgsPostgresProvider::QgsPostgresProvider( QString const &uri, const ProviderOpti
                   << QgsVectorDataProvider::NativeType( tr( "Array of number (double)" ), QStringLiteral( "double precision[]" ), QVariant::List, -1, -1, -1, -1, QVariant::Double )
                   << QgsVectorDataProvider::NativeType( tr( "Array of text" ), QStringLiteral( "text[]" ), QVariant::StringList, -1, -1, -1, -1, QVariant::String )
+                  << QgsVectorDataProvider::NativeType( tr( "Map (json)" ), QStringLiteral( "json" ), QVariant::Map, -1, -1, -1, -1, QVariant::String )
+                  << QgsVectorDataProvider::NativeType( tr( "Map (jsonb)" ), QStringLiteral( "jsonb" ), QVariant::Map, -1, -1, -1, -1, QVariant::String )
```
Jsonb is postgres > 9.4 only, right? I think this may need to be version dependant.

signedav replied:
Good input. And `json` > 9.2...

signedav committed:
pg version check on json

and comparison with == to make it equal to the other checks and avoid confusions

signedav commented:
Is it okay for you @nyalldawson ?

nyalldawson commented:
If it's ok by @m-kuhn , it's ok by me :) 

Plus you've already addressed all my comments.

luipir commented:
@signedav Ami I wrong but this only support super simple json, e.g. not nested! at least at editing level inside qgis.

signedav commented:
You can edit the json fields with a textedit widget. So it can be more complex. 
But there is not yet kind of a json-edit-widget where you could create and edit json with graphical elements.

luipir commented:
Ok, so a project should setup a specific textedit for a json column to allow edit it?
BTW I verified that editing the first level of the json does not remove all the json, it's just not visible in default editing widget.

signedav commented:
I think I remember correct that we decided not to take the textedit widget as default widget, because mainly advanced users want to edit json as text. So they have to activate it on purpose. The common use case is that the user creates a new json field to have a key/map or a list.

> BTW I verified that editing the first level of the json does not remove all the json, it's just not visible in default editing widget.

Please precise. The first level of a nested json? In the text edit widget?

andreasneumann commented:
It would be nice to have, as an alternative, a more generic JSON editor in the future - something like

https://jsoneditoronline.org/ (javascript)

or qt based:
https://github.com/peteristhegreat/qt-json-editor
https://github.com/dridk/QJsonmodel

See also 
https://stackoverflow.com/questions/998832/gui-based-or-web-based-json-editor-that-works-like-property-explorer

luipir commented:
> Please precise. The first level of a nested json? In the text edit widget?

all works correctly, but only fist json level is visible and editable using standard widget (not text widget)
